### ISSUE TITLE
Unified alerting - excessive DB calls on flattened multi-dimensional rules

### ISSUE THREAD

**Issue by wiardvanrij** [grafana/grafana#46976]
We can now leverage the unified alerting feature for alerting on multi-dimensional data. This is great. However it does create a magnitude more alerting "matching instances". To explain this, let's move to how to reproduce this

**How to reproduce it (as minimally and precisely as possible)**:

Have a Prometheus datasource that as a bunch of targets. It does not have to be a huge amount but when doing `count(up{})` it's nice to have around 200++ as result. 

Just for the sake of this issue/report, we then go create a new alert like so: 
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/5786097/160213759-c8c42bfc-e27f-45a1-a643-f5a068867bd5.png">

- Expr A: standard query with `up{}`
- Expr B: Reduce function with whatever, like avg or min is fine
- Expr C: Math, `$B == 10` (this won't trigger)
- Set evaluation period to 30s or so

**What happened**:

DB queries go brrr. 
```
k logs grafana -c grafana --since=1m | grep "INSERT INTO alert_instance" | wc -l
460
```
(460 inserts per minute = 7.6 per second)

**What you expected to happen**:

What I would expect to happen is that Grafana would evaluate this "as is" and keep it like that. IF there must be some state preserved, it should be done in a manner that is cheap. For example;
- Keep state in-memory 
- Only store state in DB if state is different per evaluation
- Perhaps only store state on the higher level and not per 'matching instance' value 

**Anything else we need to know?**:

What I've typed here is based on a first glance at the code. I believe we evaluate and init the save here: https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/schedule/schedule.go#L571
Which then loops over the states: https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/schedule/schedule.go#L653
And continuous towards the save: https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/schedule/schedule.go#L663
which eventually hits: https://github.com/grafana/grafana/blob/main/pkg/services/ngalert/store/instance_database.go#L111 which is exactly the queries I see in the query log. 

**Environment**:
- Grafana version: 8.3.3
- Data source type & version: Prometheus
- OS Grafana is installed on: Kubernetes

**Inb4 don't make these alert**
Obviously, I can make more efficient alerts (this was just an example) but the key feature of unified alerting is actually doing, what I'm doing here. As extra argument, if you run Grafana in an organisation and you allow your users to create alerts, 'oopsies' happen. It would not be nice that a user can cause a very high amount of req/s to our DB.

**Comment by armandgrillet** [comment]
This should be fixed once https://github.com/grafana/grafana/pull/52305 is merged.

**Comment by yuri-tceretian** [comment]
There is a POC https://github.com/grafana/grafana/pull/56336 that proposes to not store the Normal state to the database, and even not show it to the user (the same way Prometheus ruler works).

**Comment by yuri-tceretian** [comment]
@wiardvanrij we have done some improvements related to this problem:
- https://github.com/grafana/grafana/pull/55350 that makes alerting persist state in batches. We had to put it behind a feature flag `alertingBigTransactions`. It is available since 9.2.0 
- https://github.com/grafana/grafana/issues/61201 implements a more radical approach and makes alerting to not maintain the `Normal` state. The dimensions that stay in a normal state will not even be saved to the database, and not be displayed in UI as well. If you have alert instances in a mostly Normal state, this will greatly reduce the number of queries to the database.  It will be available in 9.4.0 and will require a feature flag `alertingNoNormalState` to be set. This is intended to be the default in Grafana 10. 
- https://github.com/grafana/grafana/pull/56575 is not directly related to your problem but also decreases the number of queries to the database.
